src/rdkafka_msg.c

```diff
@@ -1183,13 +1208,34 @@ int rd_kafka_msg_partitioner (rd_kafka_topic_t *rkt, rd_kafka_msg_t *rkm,
 
                 /* Partition not assigned, run partitioner. */
                 if (rkm->rkm_partition == RD_KAFKA_PARTITION_UA) {
-                        partition = rkt->rkt_conf.
-                                partitioner(rkt,
-                                            rkm->rkm_key,
-					    rkm->rkm_key_len,
-                                            rkt->rkt_partition_cnt,
-                                            rkt->rkt_conf.opaque,
-                                            rkm->rkm_opaque);
+
+                        if (rkt->rkt_rk->rk_conf.sticky_partition_linger_ms > 0 &&
+                            (!rkm->rkm_key || (rkm->rkm_key_len == 0 && 
+                            rkt->rkt_conf.partitioner == 
```

Comment:
Seems like we could do all this partitioner checking once on rkt initialization instead of on each message.
e.g.,
`rkt->rkt_random_partitioner = all_these_checks`